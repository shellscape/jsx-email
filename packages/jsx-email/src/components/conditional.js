import { jsx as _jsx } from "react/jsx-runtime";
export const Conditional = (props) => {
    const { children, expression, mso, head } = props;
    if (typeof expression === 'undefined' && typeof mso === 'undefined')
        throw new RangeError('jsx-email: Conditional expects the `expression` or `mso` prop to be defined');
    if (typeof expression !== 'undefined' && typeof mso !== 'undefined')
        throw new RangeError('jsx-email: Conditional expects the `expression` or `mso` prop to be defined, not both');
    // Always render a JSX custom element with data-* markers.
    // A rehype plugin will replace this element with proper conditional comments.
    // @ts-ignore - lower-case custom element tag is valid
    return (_jsx("jsx-email-cond", { "data-mso": mso, "data-expression": expression, "data-head": head, children: children }));
};
Conditional.displayName = 'Conditional';
//# sourceMappingURL=conditional.js.map