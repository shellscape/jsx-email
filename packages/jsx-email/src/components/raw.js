import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { escapeForRawComponent } from '../renderer/raw.js';
export const Raw = (props) => (_jsx(_Fragment, { children: _jsx("jsx-email-raw", { "data-skip": props.disablePlainTextOutput ? 'true' : void 0, dangerouslySetInnerHTML: { __html: `<!--${escapeForRawComponent(props.content)}-->` } }) }));
Raw.displayName = 'Raw';
//# sourceMappingURL=raw.js.map