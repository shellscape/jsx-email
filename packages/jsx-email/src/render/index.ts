export * from './render';
export * from './jsx-to-string';
export * from './process';
