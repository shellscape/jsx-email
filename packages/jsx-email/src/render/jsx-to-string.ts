/**
 * Note: Parts of this file are derived from [Hyperons](https://github.com/i-like-robots/hyperons).
 * @license MIT
 */

import type { FC, ReactNode } from 'react';
import hash from 'hash-it';

import { escapeString } from './escape-string';
import { stringifyStyles } from './stringify-styles';

const ATTR_ALIASES = {
  acceptCharset: 'acceptcharset',
  accessKey: 'accesskey',
  allowFullScreen: 'allowfullscreen',
  autoCapitalize: 'autocapitalize',
  autoComplete: 'autocomplete',
  autoCorrect: 'autocorrect',
  autoFocus: 'autofocus',
  autoPlay: 'autoplay',
  charSet: 'charset',
  className: 'class',
  colSpan: 'colspan',
  contentEditable: 'contenteditable',
  crossOrigin: 'crossorigin',
  dateTime: 'datetime',
  defaultChecked: 'checked',
  defaultSelected: 'selected',
  defaultValue: 'value',
  htmlFor: 'for',
  httpEquiv: 'http-equiv',
  longDesc: 'longdesc',
  maxLength: 'maxlength',
  minLength: 'minlength',
  noModule: 'nomodule',
  noValidate: 'novalidate',
  readOnly: 'readonly',
  referrerPolicy: 'referrerpolicy',
  rowSpan: 'rowspan',
  spellCheck: 'spellcheck',
  tabIndex: 'tabindex',
  useMap: 'usemap'
};

// <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>
const BOOLEAN_ATTRS = new Set([
  'async',
  'allowfullscreen',
  'allowpaymentrequest',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'defer',
  'disabled',
  'formnovalidate',
  'hidden',
  'ismap',
  'multiple',
  'muted',
  'novalidate',
  'nowrap',
  'open',
  'readonly',
  'required',
  'reversed',
  'selected'
]);

// https://www.w3.org/TR/html/syntax.html#void-elements
const VOID_ELEMENTS = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr'
]);

const EMPTY_OBJECT = Object.freeze({});

const promiseMap = new Map();

const wrapPromise = <TPromise extends Promise<any>>(promise: TPromise) => {
  let status = 'pending';
  let result: Awaited<TPromise>;
  const suspender = promise.then(
    (r) => {
      status = 'success';
      result = r;
    },
    (e) => {
      status = 'error';
      result = e;
    }
  );
  return {
    // eslint-disable-next-line consistent-return
    read() {
      if (status === 'pending') {
        throw suspender;
      } else if (status === 'error') {
        throw result;
      } else if (status === 'success') {
        return result;
      }
    }
  };
};

export const useData = <TData>(props: any, cb: () => Promise<TData>): TData => {
  const key = hash(props);
  let dataPromise;
  if (promiseMap.has(key)) {
    dataPromise = promiseMap.get(key);
  } else {
    dataPromise = wrapPromise(cb());
    promiseMap.set(key, dataPromise);
  }
  return dataPromise.read();
};

const renderSuspense = async (children: ReactNode[]): ReturnType<typeof jsxToString> => {
  try {
    const result = await jsxToString(children);
    return result;
  } catch (error) {
    if (error instanceof Promise) {
      await error;
      return renderSuspense(children);
    }
    throw error;
  }
};

/**
 * Convert a JSX element to a string.
 * This is a slightly modified version of Hyperons's
 * [`renderToString`](https://github.com/i-like-robots/hyperons/blob/main/src/render-to-string.js) function.
 *
 * @param element The JSX element to convert to a string
 * @returns The string representation of the JSX element
 */
export async function jsxToString(element: ReactNode): Promise<string> {
  if (element == null) {
    return '';
  }
  if (typeof element === 'string') {
    return escapeString(element);
  } else if (typeof element === 'number') {
    return String(element);
  } else if (typeof element === 'boolean' || element == null) {
    return '';
  } else if (isIterable(element)) {
    let html = '';
    for (const child of element) {
      // eslint-disable-next-line no-await-in-loop
      html += await jsxToString(child);
    }
    return html;
  }

  if (typeof (element as { $$typeof?: symbol }).$$typeof !== 'symbol') {
    throw new Error('Unsupported JSX element:', element as any);
  }

  const { type } = element;
  const props = element.props || EMPTY_OBJECT;

  if (typeof type === 'string') {
    let html = `<${type}`;

    let innerHTML = '';

    for (const prop of Object.keys(props) as ReadonlyArray<keyof typeof props & string>) {
      const value = props[prop];

      if (prop === 'children' || prop === 'key' || prop === 'ref' || value == null) {
        // Why not use a continue statement? It's slower ¯\_(ツ)_/¯
      } else if (prop === 'class' || prop === 'className') {
        // This condition is here because it is the most common attribute
        // and short-circuiting results in a ~5% performance boost.
        html += value ? ` class="${escapeString(value)}"` : '';
      } else if (prop === 'style') {
        html += ` style="${stringifyStyles(value)}"`;
      } else if (prop === 'dangerouslySetInnerHTML') {
        // eslint-disable-next-line no-underscore-dangle
        innerHTML = value.__html;
      } else {
        const name = ATTR_ALIASES[prop as keyof typeof ATTR_ALIASES] || prop;

        if (BOOLEAN_ATTRS.has(name)) {
          html += value ? ` ${name}` : '';
        } else if (typeof value === 'string') {
          html += ` ${name}="${escapeString(value)}"`;
        } else if (typeof value === 'number') {
          html += ` ${name}="${String(value)}"`;
        } else if (typeof value === 'boolean') {
          html += ` ${name}="${value}"`;
        }
      }
    }

    if (VOID_ELEMENTS.has(type)) {
      html += '/>';
    } else {
      html += '>';

      if (innerHTML) {
        html += innerHTML;
      } else {
        html += await jsxToString(props.children);
      }

      html += `</${type}>`;
    }

    return html;
  } else if (type) {
    if (typeof type === 'function') {
      return jsxToString((type as FC)(props));
    } else if (typeof type === 'symbol') {
      const key = Symbol.keyFor(type);
      // is this react fragment?
      if (key === 'react.fragment') {
        return jsxToString(props.children);
      } else if (key === 'react.suspense') {
        const suspenseResult = await renderSuspense(props.children);
        return suspenseResult;
      }
    } else if (isReactForwardRef(type)) {
      return jsxToString(
        (type as { render: (props: unknown, ref: unknown) => ReactNode }).render(props, props.ref)
      );
    }

    throw new Error(`Unsupported JSX element type: ${String(type)}`);
  }
  return '';
}

function isIterable(node: unknown): node is Iterable<unknown> {
  return typeof node === 'object' && node !== null && Symbol.iterator in node;
}

function isReactForwardRef(
  type: unknown
): type is { $$typeof: symbol; render: (props: unknown, ref: unknown) => ReactNode } {
  return (
    typeof type === 'object' &&
    type !== null &&
    'render' in type &&
    '$$typeof' in type &&
    typeof type.$$typeof === 'symbol' &&
    Symbol.keyFor(type.$$typeof) === 'react.forward_ref'
  );
}
