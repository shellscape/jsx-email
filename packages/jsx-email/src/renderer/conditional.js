/**
 * Returns a rehype plugin that replaces `<jsx-email-cond>` elements (from
 * the Conditional component) with conditional comment wrappers, based on the
 * `data-mso` and `data-expression` attributes.
 *
 * Mirrors the async factory pattern used by `getRawPlugin()`.
 */
export const getConditionalPlugin = async () => {
    const { visit } = await import('unist-util-visit');
    return function conditionalPlugin() {
        return function transform(tree) {
            const matches = [];
            let headEl;
            visit(tree, 'element', (node, index, parent) => {
                if (node.tagName === 'head')
                    headEl = node;
                if (!parent || typeof index !== 'number')
                    return;
                if (node.tagName !== 'jsx-email-cond')
                    return;
                matches.push({ index, node, parent });
            });
            for (const { node, parent, index } of matches) {
                const props = (node.properties || {});
                const msoProp = (props['data-mso'] ?? props.dataMso);
                const msoAttr = typeof msoProp === 'undefined' ? void 0 : msoProp === 'false' ? false : Boolean(msoProp);
                const exprRaw = (props['data-expression'] ?? props.dataExpression);
                const exprAttr = typeof exprRaw === 'string' ? exprRaw : void 0;
                const headProp = (props['data-head'] ?? props.dataHead);
                const toHead = typeof headProp === 'undefined'
                    ? false
                    : headProp === 'false'
                        ? false
                        : Boolean(headProp);
                let openRaw;
                let closeRaw;
                if (msoAttr === false) {
                    // Not MSO: <!--[if !mso]><!--> ... <!--<![endif]-->
                    openRaw = '<!--[if !mso]><!-->';
                    closeRaw = '<!--<![endif]-->';
                }
                else {
                    // MSO / expression path
                    const expression = exprAttr || (msoAttr === true ? 'mso' : void 0);
                    if (expression) {
                        openRaw = `<!--[if ${expression}]>`;
                        // Older Outlook/Word HTML parsers prefer the self-closing
                        // conditional terminator variant to avoid comment spillover
                        // when adjacent comments appear. Use the `<![endif]/-->` form
                        // for maximum compatibility.
                        closeRaw = '<![endif]/-->';
                    }
                }
                // If no directive attributes present, leave the element in place.
                // eslint-disable-next-line no-continue
                if (!openRaw || !closeRaw)
                    continue;
                const before = { type: 'raw', value: openRaw };
                const after = { type: 'raw', value: closeRaw };
                const children = (node.children || []);
                if (toHead && headEl) {
                    if (parent === headEl) {
                        // Replace in place: open raw, original children, close raw.
                        parent.children.splice(index, 1, before, ...children, after);
                    }
                    else {
                        // Remove wrapper from current location
                        parent.children.splice(index, 1);
                        // Append the conditional to the <head>
                        headEl.children.push(before, ...children, after);
                    }
                }
                else {
                    // Replace in place: open raw, original children, close raw.
                    parent.children.splice(index, 1, before, ...children, after);
                }
            }
        };
    };
};
//# sourceMappingURL=conditional.js.map