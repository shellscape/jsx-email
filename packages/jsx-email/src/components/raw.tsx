import React from 'react';

import type { JsxEmailComponent } from '../types.js';
import { escapeForRawComponent } from '../renderer/raw.js';

declare global {
  namespace JSX {
    interface IntrinsicElements {
      // magic custom element post-processed in render function
      'jsx-email-raw': React.DetailedHTMLProps<React.HTMLProps<HTMLElement>, HTMLElement>;
    }
  }
}

export interface RawProps {
  content: string;
  disableOutputToPlainText?: boolean;
}

export const Raw: JsxEmailComponent<RawProps> = (props) => (
  <>
    <jsx-email-raw
      data-skip={props.disableOutputToPlainText ? 'true' : undefined}
      dangerouslySetInnerHTML={{ __html: `<!--${escapeForRawComponent(props.content)}-->` }}
    />
  </>
);

Raw.displayName = 'Raw';
