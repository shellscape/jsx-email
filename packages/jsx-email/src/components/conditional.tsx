import React, { Suspense } from 'react';

import { jsxToString } from '../renderer/jsx-to-string.js';
import { useData } from '../renderer/suspense.js';
import type { JsxEmailComponent } from '../types.js';

export interface ConditionalProps {
  children?: React.ReactNode;
  expression?: string;
  head?: boolean;
  mso?: boolean;
}

const notMso = (html: string) => `<!--[if !mso]><!-->${html}<!--<![endif]-->`;

const comment = (expression: string, html: string) => `<!--[if ${expression}]>${html}<![endif]-->`;

const Renderer = (props: ConditionalProps) => {
  const { children, mso, head } = props;
  let { expression } = props;
  const html = useData(props, () => jsxToString(<>{children}</>));
  let innerHtml = '';

  if (mso === false) innerHtml = notMso(html);
  else if (mso === true && !expression) expression = 'mso';
  if (expression) innerHtml = comment(expression, html);

  const Component = head ? 'head' : 'jsx-email-cond';

  // @ts-ignore
  // Note: This is perfectly valid. TS just expects lowercase tag names to match a specific type
  return <Component dangerouslySetInnerHTML={{ __html: innerHtml }} />;
};

export const Conditional: JsxEmailComponent<ConditionalProps> = (props) => {
  const { children, expression, mso } = props;

  if (typeof expression === 'undefined' && typeof mso === 'undefined')
    throw new RangeError(
      'jsx-email: Conditional expects the `expression` or `mso` prop to be defined'
    );

  if (typeof expression !== 'undefined' && typeof mso !== 'undefined')
    throw new RangeError(
      'jsx-email: Conditional expects the `expression` or `mso` prop to be defined, not both'
    );

  return (
    <>
      <Suspense fallback={<div>waiting</div>}>
        <Renderer {...props}>{children}</Renderer>
      </Suspense>
    </>
  );
};

Conditional.displayName = 'Conditional';
