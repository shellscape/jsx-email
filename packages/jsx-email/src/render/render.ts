import { htmlToText } from 'html-to-text';

import type { PlainTextOptions, RenderOptions } from '../types';

import { jsxToString } from './jsx-to-string';
import { processHtml } from './process';

export const renderPlainText = async (
  component: React.ReactElement,
  options?: PlainTextOptions
) => {
  const result = await jsxToString(component);
  return htmlToText(result, {
    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-skip="true"]' },
      { options: { linkBrackets: false }, selector: 'a' }
    ],
    ...options
  });
};

export const render = async (component: React.ReactElement, options?: RenderOptions) => {
  const { minify = false, plainText, pretty = false } = options || {};

  if (plainText) return renderPlainText(component, typeof plainText === 'object' ? plainText : {});

  let html = await jsxToString(component);

  html = await processHtml({ html, minify, pretty });

  return html;
};
