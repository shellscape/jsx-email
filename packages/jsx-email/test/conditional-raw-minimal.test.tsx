import { jsxToString } from '../src/renderer/jsx-to-string.js';
import { render } from '../src/renderer/render.js';
import { Conditional } from '../src/components/conditional.js';
import { Raw } from '../src/components/raw.js';

// We stub core plugins to keep the bundler from trying to resolve unpublished build outputs
// during test collection. These stubs are no-ops; we're not testing plugin behavior here.
vi.mock('@jsx-email/plugin-inline', () => {
  return {
    plugin: { name: '@jsx-email/plugin-inline', symbol: Symbol.for('jsx-email/plugin') }
  };
});
vi.mock('@jsx-email/plugin-minify', () => {
  return {
    plugin: { name: '@jsx-email/plugin-minify', symbol: Symbol.for('jsx-email/plugin') }
  };
});
vi.mock('@jsx-email/plugin-pretty', () => {
  return {
    plugin: { name: '@jsx-email/plugin-pretty', symbol: Symbol.for('jsx-email/plugin') }
  };
});

const minimalFragment = (
  <Conditional mso>
    <Raw content={'<b data-testid="raw">hello</b>'} />
  </Conditional>
);

/**
 * Minimal snapshot fixture for `<Raw>` nested inside `<Conditional>`.
 *
 * This test intentionally asserts a snapshot of whatever the current
 * render pipeline produces for this structure so we can iterate on it
 * in followâ€‘ups. There are no behavioral assertions beyond the snapshot.
 */
describe('snapshot: minimal <Raw> inside <Conditional mso>', () => {
  it('renders a minimal MSO block with inlined Raw content (jsxToString)', async () => {
    const html = await jsxToString(minimalFragment);
    expect(html).toMatchSnapshot();
  });

  it('renders final HTML via the render pipeline (with plugin stubs)', async () => {
    const html = await render(minimalFragment);
    expect(html).toMatchSnapshot();
  });
});
