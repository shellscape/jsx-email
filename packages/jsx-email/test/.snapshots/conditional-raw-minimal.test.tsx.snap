// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot: minimal <Raw> inside <Conditional mso> > renders a minimal MSO block with inlined Raw content (jsxToString) 1`] = `"<jsx-email-cond><!--[if mso]><jsx-email-raw><!--<b data-testid=\\"raw\\">hello</b>--></jsx-email-raw><![endif]--></jsx-email-cond>"`;

exports[`snapshot: minimal <Raw> inside <Conditional mso> > renders final HTML via the render pipeline 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><html><body><!--[if mso]><b data-testid=\\"raw\\">hello</b><![endif]--></body></html>"`;
